lib/libspl/include/atomic.h

```diff
@@ -245,6 +245,49 @@ extern ulong_t atomic_swap_ulong(volatile ulong_t *, ulong_t);
 extern uint64_t atomic_swap_64(volatile uint64_t *, uint64_t);
 #endif
 
+/*
+ * Atomically read variable.
+ */
+#define	atomic_load_char(p)	(*(volatile uchar_t *)(p))
```

Comment:
I know libspl's versions of these functions are okay being a bit loose, but I don't believe that volatile guarantees atomicity, that's just a common implementation side-effect.
How about GCC's __atomic builtins (also provided by Clang)?

Reply:
volatile does not provide atomicity, it only makes compiler to not optimize memory accesses.  Atomicity is provided by machine register size, copied with single instruction.  That is why there are #ifdef _LP64.  This is a copy/paste from FreeBSD's atomic_common.h, used on all supported architectures, so I believe it to be correct.  I have no opinion about __atomic builtins.  When I written this, I haven't seen recent наб's commit to start using them in lib/libspl/atomic.c.  It actually makes me wonder why they were not inlined here, but that is a different topic.

Reply:
It'll be nice to finally have proper wrappers for atomic loads and stores!  This is functionality I know would be helpful elsewhere with some kstats which are a bit dodgy in this regard (though pretty harmless).

Since it sounds like these are all well tested already on FreeBSD this seems fine to me.  Though I wouldn't object to updating them to use the GCC atomic builtins for consistency in a similar fashion as was done for the other atomics.

Reply:
I am not sure what is the point now to have all the atomics in separate C file, not inlined in the header.  Is it supposed to improve compatibility when library is built with compiler supporting atomics and header is used by less capable?  Previously it had sense, since alternative implementations were written in assembler.  But now with compiler supposed to implement all of that -- what's the point?  To troubles compiler/CPU with function calls?

Reply:
Well since this is only for `libzpool.so` in user space it's probably not a huge deal since only `zdb` and `ztest` would use these (at least until there's FUSE or BUSE implementation).  The only reason I can think of to not move everything to the header (to be inlined) would be if we wanted a single shared user/kernel space header. But we don't have that today and I don't think we'd gain much from it.  We're already inlining these on the kernel side, doing the same in user space would make sense.